### PR TITLE
Add a check for certified images in the CSV

### PR DESCRIPTION
**Pull request by bcrochet**
The only reason this check should fail is if the CSV is malformed,
or the call to Pyxis fails.

This is utilizing some CSV extraction code from the operator-framework
project[1].

[1] https://github.com/operator-framework/operator-manifest-tools

Fixes: #739 

Signed-off-by: Brad P. Crochet <brad@redhat.com>

### CODE REVIEW

File: certification/internal/bundle/bundle.go [hunk 1]
```diff
@@ -166,10 +166,10 @@ func ExtractAnnotationsBytes(ctx context.Context, annotationBytes []byte) (map[s
 	return bundleMeta.Annotations, nil
 }
 
-func GetCsvFilePathFromBundle(mountedDir string) (string, error) {
+func GetCsvFilePathFromBundle(imageDir string) (string, error) {
 	log.Trace("reading clusterserviceversion file from the bundle")
-	log.Debug("mounted directory is ", mountedDir)
```
Comment (komish) [inline certification/internal/bundle/bundle.go:171]: Wow! Is this a remnant from the unshare days?

File: certification/internal/policy/operator/certified_images.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	"github.com/operator-framework/operator-manifest-tools/pkg/image"
+	"github.com/operator-framework/operator-manifest-tools/pkg/pullspec"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification/pyxis"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type imageFinder interface {
+	FindImagesByDigest(context.Context, []string) ([]pyxis.CertImage, error)
```
Comment (komish) [inline certification/internal/policy/operator/certified_images.go:17]: Needs doc

File: certification/internal/policy/operator/certified_images.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+type certifiedImagesCheck struct {
+	ImageFinder        imageFinder
+	nonCertifiedImages []string
+}
+
+func NewCertifiedImagesCheck(imageFinder imageFinder) *certifiedImagesCheck {
+	return &certifiedImagesCheck{
+		ImageFinder: imageFinder,
+	}
+}
```
Comment (komish) [inline certification/internal/policy/operator/certified_images.go:30]: My need to shuffle the doc here a bit? Documentation on the `New...` func should get picked up by godoc... but also this is an `internal` package so I'm not sure if that's true.

File: certification/internal/policy/operator/certified_images.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	log "github.com/sirupsen/logrus"
+)
+
+type imageFinder interface {
+	FindImagesByDigest(context.Context, []string) ([]pyxis.CertImage, error)
+}
+
+// CertifiedImagesCheck scans the CSV and validates that all refereenced images are certified.
+type certifiedImagesCheck struct {
+	ImageFinder        imageFinder
```
Comment (komish) [inline certification/internal/policy/operator/certified_images.go:22]: Does `imageFinder` need to be named here?

File: certification/internal/policy/operator/certified_images.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+		if !img.Certified {
+			fullImg := fmt.Sprintf("%s/%s@%s", img.Repositories[0].Registry, img.Repositories[0].Repository, img.DockerImageDigest)
+			p.nonCertifiedImages = append(p.nonCertifiedImages, fullImg)
+		}
+	}
+	return true, nil
+}
+
+func (p *certifiedImagesCheck) Name() string {
+	return "CertifiedImages"
```
Comment (komish) [inline certification/internal/policy/operator/certified_images.go:81]: I think this needs to be more descriptive. "BundleImageReferencesAreCertified", "AllImagesAreCertified" or something to that extent.

**Review by komish — CHANGES_REQUESTED** [review]
Couple of changes. Also, don't we need to adjust the way results are written before we merge this? Wasn't this check supposed to be non-blocking for now?

File: certification/internal/bundle/bundle.go (hunk 1 above)
Comment (bcrochet) [reply certification/internal/bundle/bundle.go:171]: Sure is.

File: certification/internal/policy/operator/certified_images.go (hunk 4 above)
Comment (bcrochet) [reply certification/internal/policy/operator/certified_images.go:22]: That's a good question. I can try it without it being exported.

File: certification/internal/policy/operator/certified_images.go (hunk 5 above)
Comment (bcrochet) [reply certification/internal/policy/operator/certified_images.go:81]: Done.

File: certification/internal/policy/operator/certified_images.go (hunk 4 above)
Comment (bcrochet) [reply certification/internal/policy/operator/certified_images.go:22]: Apparently not!

File: certification/internal/policy/operator/certified_images.go (hunk 3 above)
Comment (bcrochet) [reply certification/internal/policy/operator/certified_images.go:30]: Done.

File: certification/internal/policy/operator/certified_images.go (hunk 2 above)
Comment (bcrochet) [reply certification/internal/policy/operator/certified_images.go:17]: Done.

**Comment by bcrochet** [comment]
> Couple of changes. Also, don't we need to adjust the way results are written before we merge this? Wasn't this check supposed to be non-blocking for now?

As of now, it should not fail unless there is an actual error.

File: certification/internal/policy/operator/certified_images.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+func (p *certifiedImagesCheck) Name() string {
+	return "BundleImageRefsAreCertified"
+}
+
+func (p *certifiedImagesCheck) Metadata() certification.Metadata {
+	return certification.Metadata{
+		Description:      "Checking that all images referenced in the CSV are certified",
+		Level:            "optional",
+		KnowledgeBaseURL: "https://sdk.operatorframework.io/docs/olm-integration/testing-deployment/",
```
Comment (acornett21) [inline certification/internal/policy/operator/certified_images.go:94]: Shouldn't these point to our policy docs?

Comment (bcrochet) [reply certification/internal/policy/operator/certified_images.go:94]: I would say so.

**Review by acornett21 — APPROVED** [review]

**Review by komish — APPROVED** [review]
/lgtm

**Review by skattoju — APPROVED** [review]

**Pull request merged**